### ISSUE TITLE
Development Anti-Patterns in this repository as indicated by Cline+claude-opus-4.6-fast

### ISSUE THREAD

**Issue by agorevski** [ace-step/ACE-Step-1.5#504]
Repository Timestamp: `Thu Feb 12 14:38:17 2026 -0800`
Agent + Model: `Cline + vscode-lm:copilot/claude-opus-4.6-fast`
Prompt: `Please review the code in this repository for development anti-patterns and list them all out in a file called 'ANTI-PATTERNS.md'`

At the bottom you will find a table with a complete list and their severities.  Take some of these with a grain of salt - some of them are likely not really relevant, whilst others will make a lot of sense for cleanup

# Development Anti-Patterns

This document catalogs development anti-patterns identified across the ACE-Step-1.5 codebase. Each entry includes a description of the pattern, why it is problematic, and the files where it was observed.

---

## 1. God Objects / Monolithic Files

**Description:** Several files far exceed the 200 LOC hard cap defined in `AGENTS.md` and bundle multiple unrelated responsibilities into a single module or class.

**Why it matters:** Large files are harder to read, test, review, and maintain. They increase merge-conflict risk and make it difficult to reason about change impact.

**Observed in:**
- `acestep/handler.py` — orchestrates model loading, device detection, inference, audio post-processing, LoRA management, and retry logic in a single class.
- `acestep/inference.py` — mixes CLI argument parsing, model initialization, generation, and audio file I/O.
- `acestep/api_server.py` — combines FastAPI app setup, route handlers, job management, signal handling, and business logic.
- `acestep/training/trainer.py` — large training orchestrator module.
- `acestep/gradio_ui/interfaces/generation.py` — massive UI layout definition.
- `acestep/acestep_v15_pipeline.py` — pipeline class with many mixed concerns.

---

## 2. Bare `except` Clauses

**Description:** Many `try/except` blocks catch `Exception` (or use bare `except:`) without distinguishing error types, silently swallowing errors including `KeyboardInterrupt` and `SystemExit`.

**Why it matters:** Bare excepts hide bugs, make debugging extremely difficult, and can prevent graceful shutdown. They violate the `AGENTS.md` rule: *"Handle errors explicitly; avoid bare `except`."*

**Observed in:**
- `acestep/handler.py` — multiple `except Exception as e` blocks that log and continue.
- `acestep/api_server.py` — route handlers catch broad exceptions.
- `acestep/inference.py` — wraps large blocks in `try/except Exception`.
- `acestep/llm_inference.py` — catches `Exception` around LLM calls.
- `acestep/gpu_config.py` — catches `Exception` during GPU detection.
- `acestep/model_downloader.py` — catches `Exception` during downloads.
- `acestep/openrouter_adapter.py` — catches `Exception` on HTTP calls.

---

## 3. Print Debugging in Production Code

**Description:** `print()` statements are used throughout the codebase for runtime output instead of the `logging` module.

**Why it matters:** Print statements cannot be filtered by log level, cannot be routed to files or monitoring systems, and clutter stdout. `AGENTS.md` states: *"Keep logging actionable; avoid noisy logs and `print` debugging in committed code."*

**Observed in:**
- `acestep/handler.py` — `print()` for model loading status, device info, and error messages.
- `acestep/inference.py` — `print()` for progress and results.
- `acestep/api_server.py` — `print()` for server status and errors.
- `acestep/gpu_config.py` — `print()` for GPU detection output.
- `acestep/model_downloader.py` — `print()` for download progress.
- `acestep/training/trainer.py` — `print()` for training status.
- `acestep/local_cache.py` — `print()` for cache operations.

---

## 4. Mixed Responsibilities / Single Responsibility Principle Violations

**Description:** Individual functions and classes handle multiple unrelated concerns, making them difficult to test in isolation and reason about.

**Why it matters:** Violating SRP increases coupling, makes unit testing difficult, and means changes for one concern risk breaking another.

**Observed in:**
- `acestep/handler.py` — `ACEStepHandler` manages model loading, device selection, inference orchestration, audio concatenation, retry logic, and LoRA configuration.
- `acestep/api_server.py` — route handlers perform request validation, business logic, job tracking, and response formatting in single functions.
- `acestep/inference.py` — `main()` handles argument parsing, model initialization, generation, and file output.
- `acestep/gradio_ui/events/generation_handlers.py` — event handlers mix UI state management with generation orchestration.

---

## 5. Duplicated Code

**Description:** The same logic appears in multiple locations with minor variations, rather than being extracted into shared utilities.

**Why it matters:** Duplicated code means bugs must be fixed in multiple places, and divergent copies lead to subtle inconsistencies.

**Observed in:**
- **Device detection logic** duplicated across `acestep/handler.py`, `acestep/inference.py`, and `acestep/gpu_config.py` (each re-implements CUDA/MPS/XPU/CPU fallback).
- **Audio duration and token computation** duplicated between `acestep/handler.py` and `acestep/llm_inference.py`.
- **Error response formatting** repeated verbatim across multiple API route handlers in `acestep/api_server.py`.
- **Model path resolution** duplicated across `acestep/handler.py`, `acestep/model_downloader.py`, and `acestep/inference.py`.

---

## 6. Magic Numbers and Hardcoded Values

**Description:** Numeric literals with no explanation or named constant appear throughout the code.

**Why it matters:** Magic numbers make code unreadable, error-prone when the same value must be changed in multiple places, and difficult to audit.

**Observed in:**
- `32000` (sample rate) used as a literal in multiple files instead of referencing a constant.
- Duration limits (`30`, `240`) hardcoded in handler and UI code.
- Guidance scale defaults like `0.15` embedded in function calls.
- Retry counts (`10`) hardcoded in loops.
- Token computation formulas using unexplained numeric constants (e.g., `math.ceil(duration * 32000 / 128 / 4)`).

**Files:** `acestep/handler.py`, `acestep/inference.py`, `acestep/llm_inference.py`, `acestep/api_server.py`, `acestep/gradio_ui/events/generation_handlers.py`.

---

## 7. Missing or Incomplete Docstrings

**Description:** Many public functions, classes, and modules lack docstrings entirely or have only minimal descriptions.

**Why it matters:** `AGENTS.md` mandates: *"Docstrings are mandatory for all new or modified Python modules, classes, and functions."* Missing docstrings slow onboarding and make the API surface ambiguous.

**Observed in:**
- `acestep/handler.py` — `ACEStepHandler` methods lack docstrings.
- `acestep/acestep_v15_pipeline.py` — pipeline class and methods have minimal documentation.
- `acestep/api_server.py` — route handler functions lack docstrings.
- `acestep/gpu_config.py` — utility functions undocumented.
- `acestep/audio_utils.py` — audio processing helpers undocumented.
- `acestep/training/trainer.py` — training functions lack descriptions.

---

## 8. Tight Coupling Between Layers

**Description:** Higher-level modules (UI, API) directly instantiate and call into lower-level implementation classes with no abstraction boundary.

**Why it matters:** Tight coupling means changes to internals break consumers, unit testing requires the full stack, and swapping implementations is impossible.

**Observed in:**
- `acestep/gradio_ui/events/generation_handlers.py` — directly instantiates and calls `ACEStepHandler`.
- `acestep/api_server.py` — directly imports and calls handler internals.
- `acestep/inference.py` — directly constructs pipeline objects with no factory or configuration layer.
- No dependency injection or interface abstractions anywhere in the codebase.

---

## 9. Mutable Default Arguments

**Description:** Functions use mutable objects (lists, dicts) as default parameter values.

**Why it matters:** In Python, mutable default arguments are shared across all calls to the function, leading to subtle bugs where state leaks between invocations.

**Observed in:**
- Several function signatures across `acestep/handler.py` and `acestep/inference.py` use `dict` or `list` defaults.

---

## 10. Global Mutable State

**Description:** Module-level mutable variables are used to store application state, shared across the entire process.

**Why it matters:** Global mutable state creates hidden dependencies, makes testing difficult, enables race conditions in concurrent contexts, and violates `AGENTS.md`: *"Avoid hidden state and unintended side effects."*

**Observed in:**
- `acestep/api_server.py` — module-level globals `_active_jobs`, `_handler`, `_server_ref` for state management.
- `acestep/handler.py` — class-level state mixes configuration with runtime state (loaded models, device references).

---

## 11. Inconsistent Error Handling Contracts

**Description:** Functions use different strategies for reporting errors — some return `None`, some raise exceptions, some return error strings, some return tuples with status codes.

**Why it matters:** Callers cannot reliably determine how to handle failures, leading to unchecked error returns and swallowed failures.

**Observed in:**
- `acestep/handler.py` — some methods return `None` on failure, others raise.
- `acestep/api_server.py` — route handlers have inconsistent error response formats (sometimes JSON with `error` key, sometimes `detail` key, sometimes plain strings).
- `acestep/llm_inference.py` — returns `None` on failure vs. raising exceptions.
- `acestep/model_downloader.py` — mixed return-value and exception-based error reporting.

---

## 12. Dead Code and Commented-Out Code

**Description:** Commented-out imports, unreachable code paths, and unused variables remain in the codebase.

**Why it matters:** Dead code increases cognitive load, creates confusion about intent, and can mask bugs when someone uncomments stale code.

**Observed in:**
- Commented-out import statements in several modules.
- Unused variables assigned but never read.
- Conditional branches that are effectively unreachable.

---

## 13. Overly Long Functions

**Description:** Individual functions exceed 50+ lines with deep nesting (multiple levels of `if`/`try`/`for`).

**Why it matters:** Long functions are hard to understand, test, and maintain. Deep nesting increases cyclomatic complexity and bug risk.

**Observed in:**
- `acestep/handler.py` — `generate_music()` and model loading methods.
- `acestep/api_server.py` — route handler functions.
- `acestep/inference.py` — `main()` function.
- `acestep/training/trainer.py` — training loop functions.

---

## 14. String-Typed Configuration / Lack of Structured Config

**Description:** Configuration is passed as loose dictionaries and raw strings rather than typed dataclasses, enums, or Pydantic models. No validation occurs at boundaries.

**Why it matters:** Typos in dictionary keys become silent bugs, IDE support (autocomplete, type checking) is lost, and invalid configurations are only caught at runtime deep in the call stack.

**Observed in:**
- `acestep/handler.py` — configuration passed as `dict` with string keys.
- `acestep/api_server.py` — request parameters parsed from raw JSON dicts.
- `acestep/inference.py` — CLI args converted to loose dicts.
- `acestep/gradio_ui/events/generation_handlers.py` — UI state passed as untyped dictionaries.

---

## 15. Missing Type Hints

**Description:** Many functions lack parameter and return type annotations.

**Why it matters:** Type hints enable static analysis, improve IDE support, serve as documentation, and catch bugs before runtime. `AGENTS.md` recommends: *"Add type hints for new/modified functions when practical."*

**Observed in:**
- `acestep/handler.py` — most methods lack type annotations.
- `acestep/inference.py` — function signatures untyped.
- `acestep/api_server.py` — route handlers lack explicit types beyond FastAPI decorators.
- `acestep/audio_utils.py` — utility functions untyped.
- `acestep/gpu_config.py` — detection functions untyped.

---

## 16. Premature / Overbroad Warning Suppression

**Description:** `warnings.filterwarnings("ignore")` is called broadly, suppressing all warnings.

**Why it matters:** Blanket warning suppression hides deprecation notices, resource warnings, and other important signals that indicate future breakage or resource leaks.

**Observed in:**
- `acestep/inference.py` — suppresses all warnings at module level.
- `acestep/handler.py` — suppresses warnings during model loading.

---

## 17. Environment Variable Side Effects at Import Time

**Description:** Modules mutate `os.environ` at the top level (e.g., setting `TOKENIZERS_PARALLELISM`, `OMP_NUM_THREADS`, `PYTORCH_ENABLE_MPS_FALLBACK`) as a side effect of being imported.

**Why it matters:** Import-time side effects make behavior dependent on import order, break test isolation, and can silently affect unrelated parts of the application.

**Observed in:**
- `acestep/inference.py` — sets multiple environment variables at module level.
- `acestep/handler.py` — sets environment variables at module level.
- `acestep/gpu_config.py` — sets GPU-related environment variables at import time.

---

## 18. Synchronous Blocking in Async Context

**Description:** `acestep/api_server.py` uses `asyncio.to_thread()` to run synchronous handler code, but the underlying handler is not designed for thread safety or concurrent access.

**Why it matters:** Running non-thread-safe code via `asyncio.to_thread` can cause race conditions, data corruption, and GPU resource conflicts when multiple requests arrive concurrently.

**Observed in:**
- `acestep/api_server.py` — wraps handler calls in `asyncio.to_thread` without synchronization primitives.

---

## Summary

| # | Anti-Pattern | Severity | Scope |
|---|---|---|---|
| 1 | God Objects / Monolithic Files | High | Codebase-wide |
| 2 | Bare `except` Clauses | High | Codebase-wide |
| 3 | Print Debugging | Medium | Codebase-wide |
| 4 | SRP Violations | High | Core modules |
| 5 | Duplicated Code | Medium | Core modules |
| 6 | Magic Numbers | Medium | Core modules |
| 7 | Missing Docstrings | Medium | Codebase-wide |
| 8 | Tight Coupling | High | UI/API/Core |
| 9 | Mutable Default Arguments | Medium | handler, inference |
| 10 | Global Mutable State | High | api_server, handler |
| 11 | Inconsistent Error Handling | High | Codebase-wide |
| 12 | Dead / Commented-Out Code | Low | Various |
| 13 | Overly Long Functions | Medium | Core modules |
| 14 | Untyped Configuration | Medium | Codebase-wide |
| 15 | Missing Type Hints | Medium | Codebase-wide |
| 16 | Overbroad Warning Suppression | Low | inference, handler |
| 17 | Import-Time Side Effects | Medium | inference, handler, gpu_config |
| 18 | Sync Blocking in Async Context | High | api_server |

**Comment by ChuxiJ** [comment]
The project is undergoing rapid development and refactoring. Thank you for analyzing all the issues with AI—most of them stem from the AI itself. If you have a good refactoring plan, you're welcome to submit a PR to resolve them.

**Comment by schneidergithub** [comment]
This could be a case where agents.md is updated instead and most of the comments can be dismissed.  I think a triage strategy could help and start tagging issues based on priority.

**Comment by schneidergithub** [comment]
@agorevski - this is 18 issues bundled in one, it would be helpful to narrow down to 1 or 2 high priorities, if you can't fix them yourself.   Otherwise, these types of requests are going to overwhelm the project.

**Comment by DarkAlchy** [comment]
I had GLM 5 look over your code to confirm some of my suspicions, and this is what we found together.

<img width="575" height="433" alt="Image" src="https://github.com/user-attachments/assets/f7818f74-054d-4950-a131-0c0d7475abeb" />

<img width="391" height="113" alt="Image" src="https://github.com/user-attachments/assets/2b8e7e86-aff8-4540-a945-3eac13ac4a23" />

Hopefully this helps as well.

**Comment by 1larity** [comment]
Just reiterates the work indicated in #408, which is in progress. And actually done by an experienced human /grumpymorningcomment.

**Comment by merlinwarage** [comment]
Suddenly everyone has become so smart with AI
